### ISSUE TITLE
jlc does not invoke opt --passes=mem2reg, is this on purpose?

### ISSUE THREAD

**Issue by haved** [phate/jlm#244]
Running `jlc "--###" mytest.c -O3 -c -o mytest.o` provides the following output:
<details>
<summary>
Show output
</summary>

```
/usr/lib/llvm15/bin/clang -S -emit-llvm -o /tmp/tmp-mytest-clang-out.ll mytest.c

jlm-opt --llvm --FunctionInlining --InvariantValueRedirection --NodeReduction --DeadNodeElimination --ThetaGammaInversion --InvariantValueRedirection --DeadNodeElimination --NodePushOut --InvariantValueRedirection --DeadNodeElimination --NodeReduction --CommonNodeElimination --DeadNodeElimination --NodePullIn --InvariantValueRedirection --DeadNodeElimination --LoopUnrolling --InvariantValueRedirection -o /tmp/tmp-mytest-jlm-opt-out.ll /tmp/tmp-mytest-clang-out.ll

/usr/lib/llvm15/bin/llc -O3 --relocation-model=static -filetype=obj -o mytest.o /tmp/tmp-mytest-jlm-opt-out.ll
```
</details>

I am wondering if this pipeline shouldn't also invoke `opt --passes=mem2reg`, before invoking `jlm-opt`. This is based on the pipeline presented in the 2020 RVSDG paper.

I tried to run some of the benchmarks in polybench 4.2.1, and saw very little performance gain by invoking `jlm-opt` with the above list of passes. I then tried with mem2reg first, and saw a considerable difference.

To make `opt` perform any optimizations at all, and not skip all functions, I had to add a silly little sed script `s/optnone//` to remove the `optnone` attribute clang adds to all functions at -O0 (or no -O). There is probably a nicer way of avoiding this attribute in the first place.

Here is a small benchmark made using different combinations of pipeline steps, compiling polybench's `datamining/correlation`.
`"-O3"` refers to the set of flags provided by `jlc -O3`.

| Version            | clang | opt              | jlm-opt | Runtime   |
|--------------------|-------|------------------|---------|-----------|
| clang_O3           | -O3   |                  | N/A     | 3.64372s  |
| clang_O0           | -O0   |                  | N/A     | 10.95623s |
| jlm_opt_O3         | -O0   |                  | "-O3"   | 10.40696s |
| mem2reg_jlm_opt_O3 | -O0   | --passes=mem2reg | "-O3"   | 4.29448s  |
| mem2reg_only       | -O0   | --passes=mem2reg | N/A     | 8.12841s  |

Was the omission of `mem2reg` from `jlc` on purpose, awaiting an internal implementation of mem2reg?

<details>
<summary>
For reference, the full set of commands used to compile `mem2reg_jlm_opt_O3` was:
</summary>

```sh
/usr/lib/llvm15/bin/clang -DPOLYBENCH_TIME -c polybench-c-4.2.1-beta/./datamining/correlation/correlation.c -I polybench-c-4.2.1-beta/./datamining/correlation -I polybench-c-4.2.1-beta/utilities -S -emit-llvm -o build/correlation_mem2reg_jlm_opt_O3-clang-out.ll -O0

sed -i -E s/optnone// build/correlation_mem2reg_jlm_opt_O3-clang-out.ll

/usr/lib/llvm15/bin/opt build/correlation_mem2reg_jlm_opt_O3-clang-out.ll -S -o build/correlation_mem2reg_jlm_opt_O3-opt-out.ll --passes=mem2reg

../jlm/cmake-build-debug/tools/jlm-opt/jlm-opt build/correlation_mem2reg_jlm_opt_O3-opt-out.ll -o build/correlation_mem2reg_jlm_opt_O3-jlm-opt-out.ll --FunctionInlining --InvariantValueRedirection --NodeReduction --DeadNodeElimination --ThetaGammaInversion --InvariantValueRedirection --DeadNodeElimination --NodePushOut --InvariantValueRedirection --DeadNodeElimination --NodeReduction --CommonNodeElimination --DeadNodeElimination --NodePullIn --InvariantValueRedirection --DeadNodeElimination --LoopUnrolling --InvariantValueRedirection

/usr/lib/llvm15/bin/llc build/correlation_mem2reg_jlm_opt_O3-jlm-opt-out.ll -filetype=obj -o build/correlation_mem2reg_jlm_opt_O3-llc-out.o -O3

/usr/lib/llvm15/bin/clang -DPOLYBENCH_TIME -O3 -c polybench-c-4.2.1-beta/utilities/polybench.c -o build/polybench.o

/usr/lib/llvm15/bin/clang build/polybench.o build/correlation_mem2reg_jlm_opt_O3-llc-out.o -lm -o build/correlation_mem2reg_jlm_opt_O3 -no-pie

# RESULT: correlation_mem2reg_jlm_opt_O3: 4.29448
```
</details>

**Comment by phate** [comment]
Thanks for the comprehensive write-up and reporting.

Looking at Figure 7b in the paper, it clearly shows that opt --passes=mem2reg should be invoked before any invocation of jlm-opt or opt. Thus, I strongly believe that this is a bug. I would need to dig a little bit into the git history here to potentially find out what happened.

The optnone annotations did not exist in the clang version used in the paper. They were introduced in a later version. In contrast to clang/LLVM, jlm ignores them and just happily performs optimizations.

Hopefully, the pre-optimization of the code with mem2reg won't be necessary anymore soon. Jlm has nowadays an alias analysis pass that should be able to take care of this (along with the help of some other passes). It is just not enabled by default yet, as some more tests/work is needed.

**Comment by phate** [comment]
I just checked the git history. It took me a while to follow the bread crumbs before it dawned on me:

Jlc did not support llvm-opt with mem2reg at any point in time ever. Its first incarnation can be found [here](https://github.com/phate/jlm/commit/d77755bed82aca5a36be5b1b1df5b49b6adfc830#diff-90159361395de214cbcc7cd982f22626f35fc16fde81c51dddde7636876ad666), it was still creatively called `jlm` back then.

The pipeline setup for the paper came directly from the benchmark repository used for the paper. I had back then a little python script, [jlc.py](https://github.com/phate/polybench-jlm/commit/a83d5d6f217c28ac553e6cc0c01a4e1eac284e78#diff-e4be72eb7199b4eb10796dec3fe4fc68d9f01a1aea9fd593bb07cacd92591b20), which was used for producing the numbers for the paper. It invoked the llvm-opt mem2reg pass.

This python script has been at some point replaced with the direct invocation of jlc in this [commit](https://github.com/phate/polybench-jlm/commit/a83d5d6f217c28ac553e6cc0c01a4e1eac284e78#diff-e4be72eb7199b4eb10796dec3fe4fc68d9f01a1aea9fd593bb07cacd92591b20). I suspect that this change broke the (easy) reproducibility from the paper.

That being said, I see three options for moving forward:
1. You wait for the Steensgaard alias analysis + supporting optimizations (extended peephole optimizations for load and stores, extended dead node elimination, extended invariant value redirection, etc.) to be ready.
2. We temporarily (until Steensgaard++ is ready) add a flag to jlc that would invoke the mem2reg pass before the invocation of jlm-opt.
3. You hack your own "jlc.py" script together like I did back then.

Otherwise, performance numbers will always be horrible without a mem2reg (or equivalent) pass.
